Description: Don't call up_client_new() until actually needed. This blocks for some time whilst UPower is started, if it wasn't running already (which is the case during auto-login). This can delay the whole session from starting.
Author: Chris Coulson <chrisccoulson@googlemail.com>

Index: gnome-session-2.31.6/gnome-session/gsm-manager.c
===================================================================
--- gnome-session-2.31.6.orig/gnome-session/gsm-manager.c	2010-08-06 10:28:12.000000000 +0200
+++ gnome-session-2.31.6/gnome-session/gsm-manager.c	2010-08-06 10:32:58.275699997 +0200
@@ -982,12 +982,20 @@
 }
 
 static void
+manager_ensure_up_client (GsmManager *manager)
+{
+        if (manager->priv->up_client == NULL)
+                manager->priv->up_client = up_client_new ();
+}
+
+static void
 manager_attempt_hibernate (GsmManager *manager)
 {
         gboolean  can_hibernate;
         GError   *error;
         gboolean  ret;
 
+        manager_ensure_up_client (manager);  
         can_hibernate = up_client_get_can_hibernate (manager->priv->up_client);
         if (can_hibernate) {
 
@@ -1011,6 +1019,7 @@
         GError   *error;
         gboolean  ret;
 
+        manager_ensure_up_client (manager);  
         can_suspend = up_client_get_can_suspend (manager->priv->up_client);
         if (can_suspend) {
 
@@ -2540,7 +2549,7 @@
                           G_CALLBACK (on_presence_status_changed),
                           manager);
 
-        manager->priv->up_client = up_client_new ();
+        manager->priv->up_client = NULL;
 
         /* GConf setup */
         gconf_client_add_dir (manager->priv->gconf_client,
@@ -3099,6 +3108,7 @@
         gboolean can_suspend;
         gboolean can_hibernate;
 
+        manager_ensure_up_client (manager);  
         g_object_get (manager->priv->up_client,
                       "can-suspend", &can_suspend,
                       "can-hibernate", &can_hibernate,
